SCClassLibrary/Common/Collections/SequenceableCollection.sc

```diff
@@ -129,6 +129,82 @@ SequenceableCollection : Collection {
 		^true
 	}
 
+	editDistance { | other, compareFunc |
+		// only resort to the calculation if we have to
+		if(this === other) {
+			^0;
+		};
+
+		// comparison matrix is other.size
+		// reduce the memory footprint by ensuring
+		// other is the smaller array
+
+		if(this.size < other.size) {
+			^other.editDistance(this, compareFunc);
+		};
+
+		if(compareFunc.isNil) {
+			// call the primitive
+			^this.prEditDistance(other, compareFunc);
+		} {
+			// call the sclang implementation
+			^this.prLevenshteinDistance(other, compareFunc);
+		};
+	}
+
+	prEditDistance { | other, compareFunc |
+		_ArrayLevenshteinDistance
+		^this.prLevenshteinDistance(other, compareFunc);
+	}
+
+	prLevenshteinDistance { | other, compareFunc |
+		// This is the same algorithm as the primitive, just in
+		// sclang to allow equality
+		var matrix = Array.fill(other.size + 1, { | ind | ind; });
+		var upper, corner;
+
+		if(this.size == 0) {
+			^other.size;
+		};
+
+		if(other.size == 0) {
+			^this.size;
+		};
+
+		// use identity if not given another way to compare
+		compareFunc = compareFunc ? { | a, b | a === b; };
+
+		this.size.do({ | indX |
+			corner = indX;
+			matrix[0] = indX + 1;
+
+			other.size.do({ | indY |
+				upper = matrix[indY + 1];
+
+				if(compareFunc.value(this.at(indX), other.at(indY))) {
+					matrix[indY + 1] = corner;
+				} {
+					matrix[indY + 1] = [upper, corner, matrix[indY]].minItem + 1;
+				};
```

Comment:
the idea of an `if` statement assigning something seems really strange.

is there a particular performance benefit for this change?
it feels like idiomatic code at the expense of readability to me, but maybe I'm just not used to seeing it?

Reply:
> the idea of an if statement assigning something seems really strange.

`if` isn't a statement in SC, it's a method call, and assigning the result of a method call is pretty normal. :)

> is there a particular performance benefit for this change?

it does generate smaller code, because there is only one store instruction instead of two, but that's just a difference of 1 instruction and 3 bytes.

> but maybe I'm just not used to seeing it?

i think it's that. this is common in python and ruby too, and i see 84 instances of it in the class library already:

```
[brianheim@BrianMBP SCClassLibrary]$ rg "\w* = if\(" | wc
      84     791    7386
[brianheim@BrianMBP SCClassLibrary]$ rg "\w* = if\("
<snip>
editors/sc-ide/Debug/SuperCollider.app/Contents/Resources/SCClassLibrary/Common/Collections/SequenceableCollection.sc
559:			sz = if(sublist.isSequenceableCollection, { sublist.size }, { 1 });
637:		self = if(clip) { this.keep(this.size.trunc(clumpSize * numlists).postln)} { this };

editors/sc-ide/Debug/SuperCollider.app/Contents/Resources/SCClassLibrary/Common/Collections/Event.sc
485:						releaseTime = if(releaseTime.isNil) { 0.0 } { -1.0 - releaseTime };

editors/sc-ide/Debug/SuperCollider.app/Contents/Resources/SCClassLibrary/Common/Collections/Collection.sc
466:			var sz = if(sublist.isCollection)

editors/sc-ide/Debug/SuperCollider.app/Contents/Resources/SCClassLibrary/Common/Collections/SparseArray.sc
196:		res.defaultSize = if(default.isNil) { res.size } { array.size };

editors/sc-ide/Debug/SuperCollider.app/Contents/Resources/SCClassLibrary/Common/Control/Recorder.sc
109:		path = if(path.isNil) { this.makePath } { path.standardizePath };
128:			var doneAction = if(duration <= 0, 0, 2);

<snip>
```